<!DOCTYPE html><html lang="en"><head><title>cli/index</title></head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0"><meta name="groc-relative-root" content="../"><meta name="groc-document-path" content="cli/index"><meta name="groc-project-path" content="cli/index.js"><meta name="groc-github-url" content="https://github.com/dglmoore/gemma"><link rel="stylesheet" type="text/css" media="all" href="../assets/style.css"><script type="text/javascript" src="../assets/behavior.js"></script><body><div id="meta"><div class="file-path"><a href="https://github.com/dglmoore/gemma/blob/master/cli/index.js">cli/index.js</a></div></div><div id="document"><div class="segment"></div><div class="segment"><div class="comments "><div class="wrapper"><p>eslint-disable no-console </p>
<h1 id="gemma">gemma</h1>
<p>The <code>gemma</code> command line interface provides a (hopefully) simple way for you
to build and interact with a <strong>Gemma</strong> database.</p>
<h2 id="table-of-contents">Table of Contents</h2>
<ul>
<li><a href="#example-usage">Example Usage</a></li>
<li><a href="#global-options">Global Options</a></li>
<li><a href="#import-subcommands">Import Subcommands</a></li>
<li><a href="#querying">Querying</a></li>
<li><a href="#searching-quickgo">Searching QuickGO</a></li>
<li><a href="#epilog">Epilog</a></li>
</ul></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="example-usage">Example Usage</h2>
<pre><code class="lang-shell">λ gemma import-vcf ./data/example.vcf
…  awaiting  Importing VCF file ./data/example.vcf
☒  complete  90 VCF entries imported</code></pre>
<pre><code class="lang-shell">λ gemma import-gff ./data/example.gff
…  awaiting  Importing GFF file ./data/example.gff
☒  complete  99640 GFF entries imported</code></pre>
<pre><code class="lang-shell">λ gemma import-env ./data/example.csv
…  awaiting  Importing environment variables from ./data/example.csv
☒  complete  environment imported</code></pre>
<pre><code class="lang-shell">λ gemma list-samples
PM10
PM112
PM4
PM5
WSB16
WSB161
WSB50
WSB57
WSB77</code></pre>
<pre><code class="lang-shell">λ gemma list-chromosomes
ScCC6lQ_1;HRSCAF_16</code></pre></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">const</span> program = <span class="hljs-built_in">require</span>(<span class="hljs-string">'commander'</span>);
<span class="hljs-keyword">const</span> chalk = <span class="hljs-built_in">require</span>(<span class="hljs-string">'chalk'</span>);
<span class="hljs-keyword">const</span> signale = <span class="hljs-built_in">require</span>(<span class="hljs-string">'signale'</span>);
<span class="hljs-keyword">const</span> gemma = <span class="hljs-built_in">require</span>(<span class="hljs-string">'../lib'</span>);
<span class="hljs-keyword">const</span> util = <span class="hljs-built_in">require</span>(<span class="hljs-string">'./util'</span>);</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="global-options">Global Options</h2>
<ul>
<li><a href="#help">--help</a></li>
<li><a href="#version">--version</a></li>
<li><a href="#database-path">--database</a></li>
<li><a href="#silent-mode">--silent</a></li>
<li><a href="#no-color">--no-color</a></li>
</ul></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>All of the subcommands that <code>gemma</code> provides accept and respect number of
common options.</p></div></div><div class="code"><div class="wrapper">program</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="help">Help</h3>
<p>If you do not not provide any subcommands, <code>gemma</code> will print the help
information</p>
<pre><code class="lang-shell">λ gemma
Usage: gemma [options] [command]
Options:
  -v, --version                           output the version number
  -d, --database &lt;filename&gt;               Database filename (default: &quot;./gemma.db&quot;)
  -s, --silent                            Generate no non-essential logging
  --no-color                              Do not colorize output
  -h, --help                              output usage information
Commands:
  import-vcf|iv [options] &lt;vcf-filename&gt;  Import a VCF file into database
  import-gff|ig &lt;gff-filename&gt;            Import features from a GFF file into database
  import-env|ie [options] &lt;csv-filename&gt;  Import environmental data into database
  list-samples|lss                        List the distinct sample IDs
  list-chromosomes|lsc                    List the distinct chromosomes</code></pre>
<h3 id="version">Version</h3>
<p>At any point along the way, you can print <strong>Gemma</strong>&#39;s current version.</p>
<pre><code class="lang-shell">λ gemma -v
0.0.0</code></pre></div></div><div class="code"><div class="wrapper">    .version(util.version(), <span class="hljs-string">'-v, --version'</span>)</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="database-path">Database Path</h3>
<p>By default, <strong>Gemma</strong>&#39;s database will be constructed in the current
working directory and be named <code>gemma.db</code>. You can override this path
using the <code>-d</code> or <code>--database</code> option, e.g.</p>
<pre><code class="lang-shell">λ gemma -d data/gemma.db</code></pre>
<p>will create or use the database <code>data/gemma.db</code> file.</p></div></div><div class="code"><div class="wrapper">    .option(<span class="hljs-string">'-d, --database &lt;filename&gt;'</span>, <span class="hljs-string">'Database filename'</span>, <span class="hljs-string">'./gemma.db'</span>)</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="silent-mode">Silent Mode</h3>
<p>Most subcommands, particularly the import functions, produce extra
logging information to let you know how things are going. If you&#39;d like
to suppress that, you can provide the <code>-s</code> or <code>--silent</code> flag.</p>
<pre><code class="lang-shell">λ gemma --silent</code></pre>
<p>Only error messages and the obviously desired output will be generated.</p></div></div><div class="code"><div class="wrapper">    .option(<span class="hljs-string">'-s, --silent'</span>, <span class="hljs-string">'Generate no non-essential logging'</span>)</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="no-color">No Color</h3>
<p>We live in the 21st century which brings with it terminals that support
colorized output. We tried to make use of that as much as possible.
However, when using command line utilities from a script, you might not
want <strong>Gemma</strong>&#39;s output to have the ASCII escape codes that facilitate
coloring. To suppress it, you can provide the <code>--no-color</code> option.</p>
<pre><code class="lang-shell">λ gemma --no-color</code></pre>
<p>Alternatively, you can set the <code>FORCE_COLOR=0</code> in environment variable.</p>
<pre><code class="lang-bash">λ FORCE_COLOR=<span class="hljs-number">0</span> gemma</code></pre></div></div><div class="code"><div class="wrapper">    .option(<span class="hljs-string">'--no-color'</span>, <span class="hljs-string">'Do not colorize output'</span>);</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="import-subcommands">Import Subcommands</h2>
<ul>
<li><a href="#import-vcf">import-vcf</a></li>
<li><a href="#import-gff">import-gff</a></li>
<li><a href="#import-environment">import-env</a></li>
</ul></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="import-vcf">Import VCF</h3>
<p>Often the first step in the process of building your <strong>Gemma</strong> database will
be to import a Variant Call Format (VCF) file. This can be done with the
<code>import-vcf &lt;vcf-filename&gt;</code> subcommand.</p>
<pre><code class="lang-shell">λ gemma import-vcf data/example.vcf
…  awaiting  Importing VCF file data/example.vcf
☒  complete  90 VCF entries imported</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'import-vcf &lt;vcf-filename&gt;'</span>)
    .alias(<span class="hljs-string">'iv'</span>)
    .description(<span class="hljs-string">'Import a VCF file into database'</span>)</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>In the event that you call this command a second time on a given
database, you will get an error:</p>
<pre><code class="lang-shell">λ gemma import-vcf data/different.vcf
…  awaiting  Importing VCF file ./data/example.vcf
✖  error     VCF data has already been imported; rerun with --append option to force</code></pre>
<p>By default, <strong>Gemma</strong> will not let you import VCF data twice. If you
know that the Sample IDs, chromosomes, or positions in the VCF file you
are trying to import have not yet been imported, you can override this
block using the <code>-a</code> or <code>--append</code> flag.</p>
<pre><code class="lang-shell">λ gemma import-vcf --append data/different.vcf</code></pre>
<p><strong>NOTE</strong>: <strong>Gemma</strong> expects the sample name, chromosome name and
position of each variant to be unique. If you try to import duplicate
data with <code>--append</code>, you&#39;ll get an error:</p>
<pre><code class="lang-shell">λ gemma import-vcf --append data/example.vcf
…  awaiting  Importing VCF file ./data/example.vcf
✖  error     An entry in the VCF has already been imported into the database</code></pre></div></div><div class="code"><div class="wrapper">    .option(<span class="hljs-string">'-a, --append'</span>, <span class="hljs-string">'Append data to the database'</span>)
    .action(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">filename, cmd</span>) </span>{
        cmd = util.collectOpts(cmd, <span class="hljs-string">'append'</span>);
        cmd.silent || signale.await(<span class="hljs-string">`Importing VCF file <span class="hljs-subst">${filename}</span>`</span>);
        gemma.import.vcf(filename, cmd)
            .then(n =&gt; cmd.silent || signale.complete(<span class="hljs-string">`<span class="hljs-subst">${n}</span> VCF entries imported`</span>))
            .catch(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">err</span>) </span>{
                <span class="hljs-keyword">if</span> (err.code &amp;&amp; err.code === <span class="hljs-string">'SQLITE_CONSTRAINT'</span>) {
                    signale.error(<span class="hljs-string">'An entry in the VCF has already been imported into the database'</span>);
                } <span class="hljs-keyword">else</span> {
                    signale.error(err);
                }
            });
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="import-gff">Import GFF</h3>
<p>You can import a Generic Features Format (GFF) file using the `import-gff</p>
<p><gff-filename>` subcommand:</p>
<pre><code class="lang-shell">λ gemma import-gff ./data/example.gff
…  awaiting  Importing GFF file ./data/example.gff
☒  complete  99640 GFF entries imported</code></pre>
<p>Just as with the <a href="#import-vcf">import-vcf</a> subcommand, <code>import-gff</code> will
produce an error if you try to import more than one GFF file.</p>
<pre><code class="lang-shell">λ gemma import-gff ./data/example.gff
…  awaiting  Importing GFF file ./data/example.gff
✖  error     GFF data has already been imported; the --append option is not supported yet</code></pre>
<p>However, as the error message suggests, the <code>--append</code> option is not yet
supported for <code>import-gff</code>. We should be adding it very soon.</p></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'import-gff &lt;gff-filename&gt;'</span>)
    .alias(<span class="hljs-string">'ig'</span>)
    .description(<span class="hljs-string">'Import features from a GFF file into database'</span>)
    .action(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">filename, cmd</span>) </span>{
        cmd = util.collectOpts(cmd);
        cmd.silent || signale.await(<span class="hljs-string">`Importing GFF file <span class="hljs-subst">${filename}</span>`</span>);
        gemma.import.gff(filename, cmd)
            .then(n =&gt; cmd.silent || signale.complete(<span class="hljs-string">`<span class="hljs-subst">${n}</span> GFF entries imported`</span>))
            .catch(err =&gt; signale.error(err));
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="import-environment">Import Environment</h3>
<p>You can import metadata (e.g. GPS coordinate, altitude, temperature, etc...)
about the samples provided in your VCF file. These data <del>can</del> will be used
to preform various analyses down the road. The data should be provided in
comma-separated form (CSV) and include, at least, a <code>SampleID</code> column whose
value corresponds to a sample in your VCF.</p>
<pre><code class="lang-shell">λ gemma import-env data/example.csv
…  awaiting  Importing environment variables from data/example.csv
☒  complete  environment imported</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'import-env &lt;csv-filename&gt;'</span>)
    .alias(<span class="hljs-string">'ie'</span>)
    .description(<span class="hljs-string">'Import environmental data into database'</span>)</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Just like with <a href="#import-vcf">import-vcf</a>, if you call this command a
second time on a given database, you will get an error:</p>
<pre><code class="lang-shell">λ gemma import-env data/example.csv
…  awaiting  Importing environment variables from data/example.csv
✖  error     Environment data has already been imported; rerun with --append option to force</code></pre>
<p>The <code>-a</code> or <code>--append</code> option will allow you to import data anyway.</p>
<pre><code class="lang-shell">λ gemma import-vcf --append data/different.vcf</code></pre>
<p>Again, if you try to import duplicate data with <code>--append</code>, you&#39;ll get
an error:</p>
<pre><code class="lang-shell">λ gemma import-vcf --append data/example.vcf
…  awaiting  Importing VCF file ./data/example.vcf
✖  error     An entry in the VCF has already been imported into the database</code></pre></div></div><div class="code"><div class="wrapper">    .option(<span class="hljs-string">'-a, --append'</span>, <span class="hljs-string">'Append data to the database'</span>)
    .action(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">filename, cmd</span>) </span>{
        cmd = util.collectOpts(cmd, <span class="hljs-string">'append'</span>);
        cmd.silent || signale.await(<span class="hljs-string">`Importing environment variables from <span class="hljs-subst">${filename}</span>`</span>);
        gemma.import.env(filename, cmd)
            .then(() =&gt; cmd.silent || signale.complete(<span class="hljs-string">'environment imported'</span>))
            .catch(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">err</span>) </span>{
                <span class="hljs-keyword">if</span> (err.code &amp;&amp; err.code === <span class="hljs-string">'SQLITE_CONSTRAINT'</span>) {
                    signale.error(<span class="hljs-string">'An environment variable has already been imported into the database'</span>);
                } <span class="hljs-keyword">else</span> {
                    signale.error(err);
                }
            });
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="querying">Querying</h2>
<p>You can query an existing <strong>Gemma</strong> database in a <del>number of</del> two ways.</p>
<ul>
<li><a href="#list-samples">list-samples</a></li>
<li><a href="#list-chromosomes">list-chromosomes</a></li>
</ul></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="list-samples">List Samples</h3>
<p>This just prints the sample IDs for all of the samples in the database, each
on their own line.</p>
<pre><code class="lang-shell">λ gemma list-samples
PM10
PM112
PM4
PM5
WSB16
WSB161
WSB50
WSB57
WSB77</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'list-samples'</span>)
    .alias(<span class="hljs-string">'lss'</span>)
    .description(<span class="hljs-string">'List the distinct sample IDs'</span>)
    .action(<span class="hljs-keyword">async</span> <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">...args</span>) </span>{
        <span class="hljs-keyword">const</span> cmd = util.collectOpts(util.last(args));
        <span class="hljs-keyword">const</span> samples = <span class="hljs-keyword">await</span> gemma.query.listSamples(cmd);
        <span class="hljs-keyword">if</span> (samples.length) {
            samples.forEach(s =&gt; process.stdout.write(s + <span class="hljs-string">'\n'</span>));
        } <span class="hljs-keyword">else</span> {
            process.stdout.write(chalk.red.bold(<span class="hljs-string">'no samples found'</span>));
        }
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="list-chromosomes">List Chromosomes</h3>
<p>This just prints the chromosome names that exist in the database, each on
their own line.</p>
<pre><code class="lang-shell">λ gemma list-chromosomes
ScCC6lQ_1;HRSCAF_16</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'list-chromosomes'</span>)
    .alias(<span class="hljs-string">'lsc'</span>)
    .description(<span class="hljs-string">'List the distinct chromosomes'</span>)
    .action(<span class="hljs-keyword">async</span> <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">...args</span>) </span>{
        <span class="hljs-keyword">const</span> cmd = util.collectOpts(util.last(args));
        <span class="hljs-keyword">const</span> chromosomes = <span class="hljs-keyword">await</span> gemma.query.listChromosomes(cmd);
        <span class="hljs-keyword">if</span> (chromosomes.length) {
            chromosomes.forEach(c =&gt; process.stdout.write(c + <span class="hljs-string">'\n'</span>));
        } <span class="hljs-keyword">else</span> {
            process.stdout.write(chalk.red.bold(<span class="hljs-string">'no chromosomes found\n'</span>));
        }
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="searching-quickgo">Searching QuickGo</h2>
<p>You can search the European Bioinformatics Institute&#39;s QuickGO database using <strong>Gemma</strong>&#39;s command line interface. This is useful when you don&#39;t quite know what you&#39;re looking for yet. We provide, at the moment, two search methods:</p>
<ul>
<li><a href="#go-term">go-term</a></li>
<li><a href="#go-search">go-search</a></li>
<li><a href="#go-proteins">go-proteins</a></li>
</ul></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="go-term">GO Term</h3>
<p>This searches the QuickGo database for entries with a specific GO
term ID and prints the result to the screen.</p>
<pre><code class="lang-shell">λ gemma go-term GO:0036026
┌─────────┬──────────────┬────────────────────────────────────┬──────────┬──────────────────────┐
│ (index) │      id      │                name                │ synonyms │        aspect        │
├─────────┼──────────────┼────────────────────────────────────┼──────────┼──────────────────────┤
│    0    │ &#39;GO:0036026&#39; │ &#39;protein C inhibitor-PLAT complex&#39; │    6     │ &#39;cellular_component&#39; │
└─────────┴──────────────┴────────────────────────────────────┴──────────┴──────────────────────┘</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'go-term &lt;goterm&gt;'</span>)
    .description(<span class="hljs-string">'Search the EBI gene ontology database for a specific GO term'</span>)
    .action(<span class="hljs-keyword">async</span> <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">goterm, cmd</span>) </span>{
        gemma.query.goTerm(goterm, cmd)
            .then(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">results</span>) </span>{
                <span class="hljs-keyword">if</span> (results.length) {
                    results.forEach((result) =&gt; result.synonyms = result.synonyms.length);
                    <span class="hljs-built_in">console</span>.table(results);
                } <span class="hljs-keyword">else</span> {
                    <span class="hljs-keyword">const</span> msg = <span class="hljs-string">`no results found for GO term <span class="hljs-subst">${goterm}</span>\n`</span>;
                    process.stdout.write(chalk.red.bold(msg));
                }
            }).catch(err =&gt; signale.error(err.message));
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="go-search">GO Search</h3>
<p>This searches the QuickGo database for generic entries. You can provide
any kind of query you like. For example, you can search for <em>PLAT</em>:</p>
<pre><code class="lang-shell">λ gemma go-search PLAT
┌─────────┬──────────────┬───────────────────────────────────────────────────┬──────────────────────┐
│ (index) │      id      │                       name                        │        aspect        │
├─────────┼──────────────┼───────────────────────────────────────────────────┼──────────────────────┤
│    0    │ &#39;GO:0036026&#39; │        &#39;protein C inhibitor-PLAT complex&#39;         │ &#39;cellular_component&#39; │
│    1    │ &#39;GO:0030168&#39; │               &#39;platelet activation&#39;               │ &#39;biological_process&#39; │
│    2    │ &#39;GO:0070527&#39; │              &#39;platelet aggregation&#39;               │ &#39;biological_process&#39; │
│    3    │ &#39;GO:0030220&#39; │               &#39;platelet formation&#39;                │ &#39;biological_process&#39; │
│    4    │ &#39;GO:0002576&#39; │             &#39;platelet degranulation&#39;              │ &#39;biological_process&#39; │
│    5    │ &#39;GO:0070090&#39; │                 &#39;metaphase plate&#39;                 │ &#39;cellular_component&#39; │
│    6    │ &#39;GO:1990073&#39; │                &#39;perforation plate&#39;                │ &#39;cellular_component&#39; │
│    7    │ &#39;GO:0097218&#39; │                   &#39;sieve plate&#39;                   │ &#39;cellular_component&#39; │
│    8    │ &#39;GO:0036344&#39; │             &#39;platelet morphogenesis&#39;              │ &#39;biological_process&#39; │
│    9    │ &#39;GO:0036345&#39; │               &#39;platelet maturation&#39;               │ &#39;biological_process&#39; │
│   10    │ &#39;GO:0009504&#39; │                   &#39;cell plate&#39;                    │ &#39;cellular_component&#39; │
│   11    │ &#39;GO:0032437&#39; │                 &#39;cuticular plate&#39;                 │ &#39;cellular_component&#39; │
│   12    │ &#39;GO:0003142&#39; │         &#39;cardiogenic plate morphogenesis&#39;         │ &#39;biological_process&#39; │
│   13    │ &#39;GO:0042827&#39; │             &#39;platelet dense granule&#39;              │ &#39;cellular_component&#39; │
│   14    │ &#39;GO:0021999&#39; │ &#39;neural plate anterior/posterior regionalization&#39; │ &#39;biological_process&#39; │
│   15    │ &#39;GO:0021997&#39; │         &#39;neural plate axis specification&#39;         │ &#39;biological_process&#39; │
│   16    │ &#39;GO:0021998&#39; │    &#39;neural plate mediolateral regionalization&#39;    │ &#39;biological_process&#39; │
│   17    │ &#39;GO:0021991&#39; │             &#39;neural plate thickening&#39;             │ &#39;biological_process&#39; │
│   18    │ &#39;GO:0021990&#39; │             &#39;neural plate formation&#39;              │ &#39;biological_process&#39; │
│   19    │ &#39;GO:1990265&#39; │     &#39;platelet-derived growth factor complex&#39;      │ &#39;cellular_component&#39; │
│   20    │ &#39;GO:0070889&#39; │       &#39;platelet alpha granule organization&#39;       │ &#39;biological_process&#39; │
│   21    │ &#39;GO:0033505&#39; │            &#39;floor plate morphogenesis&#39;            │ &#39;biological_process&#39; │
│   22    │ &#39;GO:0033504&#39; │             &#39;floor plate development&#39;             │ &#39;biological_process&#39; │
│   23    │ &#39;GO:0070560&#39; │          &#39;protein secretion by platelet&#39;          │ &#39;biological_process&#39; │
│   24    │ &#39;GO:0070541&#39; │            &#39;response to platinum ion&#39;             │ &#39;biological_process&#39; │
└─────────┴──────────────┴───────────────────────────────────────────────────┴──────────────────────┘</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'go-search &lt;query&gt;'</span>)
    .description(<span class="hljs-string">'Search the EBI gene ontology database for a general query'</span>)
    .action(<span class="hljs-keyword">async</span> <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">query, cmd</span>) </span>{
        gemma.query.goSearch(query, cmd)
            .then(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">results</span>) </span>{
                <span class="hljs-keyword">if</span> (results.length) {
                    <span class="hljs-built_in">console</span>.table(results);
                } <span class="hljs-keyword">else</span> {
                    <span class="hljs-keyword">const</span> msg = <span class="hljs-string">`no results found for query <span class="hljs-subst">${query}</span>\n`</span>;
                    process.stdout.write(chalk.red.bold(msg));
                }
            }).catch(err =&gt; signale.error(err.message));
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h3 id="go-proteins">GO Proteins</h3>
<p>This searches the EBI proteins database for proteins matching a GO term.
For example, you can search for proteins associated with the &quot;Protein C
inhibitor-PLAT complex&quot; (GO:0036026)</p>
<pre><code class="lang-shell">λ gemma go-proteins GO:0036026
┌─────────┬────────────────────┬───────────────────────────────────────────────────────────────┬──────────────────┐
│ (index) │         id         │                            protein                            │      genes       │
├─────────┼────────────────────┼───────────────────────────────────────────────────────────────┼──────────────────┤
│    0    │ &#39;A0A091CUK9_FUKDA&#39; │              &#39;Plasma serine protease inhibitor&#39;               │ [ &#39;H920_17264&#39; ] │
│    1    │ &#39;A0A096NJK2_PAPAN&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│    2    │ &#39;A0A0D9REJ2_CHLSB&#39; │                  &#39;Serpin family A member 5&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│    3    │ &#39;A0A1D5QZ04_MACMU&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│    4    │ &#39;A0A1U7U5R8_TARSY&#39; │              &#39;plasma serine protease inhibitor&#39;               │  [ &#39;SERPINA5&#39; ]  │
│    5    │ &#39;A0A1U8BVH3_MESAU&#39; │              &#39;plasma serine protease inhibitor&#39;               │  [ &#39;Serpina5&#39; ]  │
│    6    │ &#39;A0A2I2YDG3_GORGO&#39; │                  &#39;Serpin family A member 5&#39;                   │    undefined     │
│    7    │ &#39;A0A2J8QMG0_PANTR&#39; │                     &#39;SERPINA5 isoform 1&#39;                      │  [ &#39;SERPINA5&#39; ]  │
│    8    │ &#39;A0A2K5BUG0_AOTNA&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│    9    │ &#39;A0A2K5JNV2_COLAP&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   10    │ &#39;A0A2K5MCI7_CERAT&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   11    │ &#39;A0A2K5Q5I2_CEBCA&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   12    │ &#39;A0A2K5WK85_MACFA&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   13    │ &#39;A0A2K5YI81_MANLE&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   14    │ &#39;A0A2K6CF53_MACNE&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   15    │ &#39;A0A2K6EKL0_PROCO&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   16    │ &#39;A0A2K6LR94_RHIBE&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   17    │ &#39;A0A2K6PIF0_RHIRO&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   18    │ &#39;A0A2K6V1R2_SAIBB&#39; │                  &#39;Serpin family A member 5&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   19    │ &#39;A0A2R9AU45_PANPA&#39; │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   20    │   &#39;D2HEM8_AILME&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   21    │   &#39;E2RMF9_CANLF&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   22    │    &#39;F1SCE3_PIG&#39;    │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   23    │    &#39;F7EMJ6_RAT&#39;    │ &#39;Serine (Or cysteine) peptidase inhibitor, clade A, member 5&#39; │  [ &#39;Serpina5&#39; ]  │
│   24    │   &#39;F7GRA1_CALJA&#39;   │                  &#39;Serpin family A member 5&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   25    │   &#39;G1S667_NOMLE&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   26    │   &#39;G1SQG6_RABIT&#39;   │                  &#39;Serpin family A member 5&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   27    │   &#39;G3T0X8_LOXAF&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   28    │   &#39;G5B492_HETGA&#39;   │              &#39;Plasma serine protease inhibitor&#39;               │ [ &#39;GW7_16185&#39; ]  │
│   29    │   &#39;H0VZ56_CAVPO&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   30    │   &#39;H0X6H4_OTOGA&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   31    │   &#39;H2NM57_PONAB&#39;   │                  &#39;Serpin family A member 5&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   32    │   &#39;I3NE99_ICTTR&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   33    │   &#39;L8ILP6_9CETA&#39;   │              &#39;Plasma serine protease inhibitor&#39;               │ [ &#39;M91_13739&#39; ]  │
│   34    │   &#39;M3WCX5_FELCA&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   35    │   &#39;M3XVL4_MUSPF&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
│   36    │    &#39;IPSP_HUMAN&#39;    │              &#39;Plasma serine protease inhibitor&#39;               │  [ &#39;SERPINA5&#39; ]  │
│   37    │    &#39;IPSP_MOUSE&#39;    │              &#39;Plasma serine protease inhibitor&#39;               │  [ &#39;Serpina5&#39; ]  │
│   38    │    &#39;IPSP_BOVIN&#39;    │              &#39;Plasma serine protease inhibitor&#39;               │  [ &#39;SERPINA5&#39; ]  │
│   39    │   &#39;W5Q0L2_SHEEP&#39;   │                   &#39;Uncharacterized protein&#39;                   │  [ &#39;SERPINA5&#39; ]  │
└─────────┴────────────────────┴───────────────────────────────────────────────────────────────┴──────────────────┘</code></pre></div></div><div class="code"><div class="wrapper">program
    .command(<span class="hljs-string">'go-proteins &lt;goterm&gt;'</span>)
    .description(<span class="hljs-string">'Search the EBI proteins database for gene products associated with a given gene ontology'</span>)
    .action(<span class="hljs-keyword">async</span> <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">query, cmd</span>) </span>{
        gemma.query.goProteins(query, cmd)
            .then(<span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">results</span>) </span>{
                <span class="hljs-keyword">if</span> (results.length) {
                    <span class="hljs-built_in">console</span>.table(results);
                } <span class="hljs-keyword">else</span> {
                    <span class="hljs-keyword">const</span> msg = <span class="hljs-string">`no results found for GO term <span class="hljs-subst">${query}</span>\n`</span>;
                    process.stdout.write(chalk.red.bold(msg));
                }
            }).catch(err =&gt; signale.error(err.message));
    });</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><h2 id="epilog">Epilog</h2></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>If you provide an unrecognized subcommand, we print the <a href="#help">help information</a>.</p></div></div><div class="code"><div class="wrapper">program.on(<span class="hljs-string">'command:*'</span>, <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params"></span>) </span>{
    process.stdout.write(program.helpInformation() + <span class="hljs-string">'\n'</span>);
    process.exit(<span class="hljs-number">1</span>);
});</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Everything up to this point just sets up the command-line interface. We
still need to tell the CLI to parse the arguments and execute the
appropriate actions.</p></div></div><div class="code"><div class="wrapper">program.parse(process.argv);</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>If you don&#39;t provide any arguments, we print the <a href="#help">help information</a>.</p></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">if</span> (process.argv.length == <span class="hljs-number">2</span>) {
    process.stderr.write(program.helpInformation() + <span class="hljs-string">'\n'</span>);
    process.exit(<span class="hljs-number">1</span>);
}</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>It is always possible for an error to slip by error handling code. To make
sure you see a nice error message, we process all unhandled promise
rejections.</p></div></div><div class="code"><div class="wrapper">process.on(<span class="hljs-string">'unhandledRejection'</span>, <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">err</span>) </span>{
    signale.error(err);
    process.exit(<span class="hljs-number">1</span>);
});</div></div></div></div></body></html>